test/task.cpp

```diff
@@ -0,0 +1,95 @@
+#include <gtest/gtest.h>
+
+#include <functional>
+#include <yaclib/task.hpp>
+
+namespace {
+
+using namespace yaclib;
+
+int counter = 0;
+void AddOne() {
+  counter++;
+}
+
+struct Idle {
+  void DoNothing() {
+    const_method_called = false;
+  }
+
+  void DoNothing() const {
+    const_method_called = true;
+  };
+
+  void DoNothingForceConst() const {
+    const_method_called = true;
+  }
+
+  bool mutable const_method_called{false};
+};
+
+GTEST_TEST(task, function) {
+  counter = 0;
+  ITaskPtr task = CreateTask(&AddOne);
+  EXPECT_EQ(counter, 0);
+  task->Call();
+  EXPECT_EQ(counter, 1);
+  task->Call();
+  EXPECT_EQ(counter, 2);
+
+  task = CreateTask(&AddOne);
+  EXPECT_EQ(counter, 2);
+  task->Call();
+  EXPECT_EQ(counter, 3);
+}
+
+GTEST_TEST(task, lambda) {
+  size_t value = 0;
+  auto lambda = [&] { ++value; };
+  ITaskPtr task = CreateTask(lambda);
+  EXPECT_EQ(value, 0);
+  task->Call();
+  EXPECT_EQ(value, 1);
+  task->Call();
+  EXPECT_EQ(value, 2);
+
+  task = CreateTask([&] { ++value; });
+  EXPECT_EQ(value, 2);
+  task->Call();
+  EXPECT_EQ(value, 3);
+}
+
+GTEST_TEST(task, std_function) {
+  int value = 0;
+  std::function<void(void)> fun = [&] { ++value; };
+
+  auto task = CreateTask(fun);
+  EXPECT_EQ(value, 0);
+  task->Call();
+  EXPECT_EQ(value, 1);
+  task->Call();
+  EXPECT_EQ(value, 2);
+
+  task = CreateTask(fun);
+  EXPECT_EQ(value, 2);
+  task->Call();
+  EXPECT_EQ(value, 3);
+}
+
+GTEST_TEST(task, member_function) {
```

Comment:
I'm actually not very sure if we need functionality that the user can trivially implement himself

Reply:
There are only two [primitive] overloads of `CreateTask(...)` so that's easy to maintain, wherein we are simplifying user's life (we have thought carefully about these overloads) for low cost.